web/ww.ts

```diff
@@ -29,82 +37,209 @@ enum WormholeErrorCodes {
-// JS expects lowercase dictionary entries.
-type GoICEServers = [{ URLs:[string], Username: string, Credential: string }];
+type State = (msg: string) => Promise<State>;
 
 class Wormhole {
 	// Signalling protocol version.
 	static readonly protocol = "4";
 
-	slot?: number;
 	pass: Uint8Array;
-	state: string; // TODO enum?
+	signalserver: string;
+	slot?: number;
 	pc?: RTCPeerConnection;
-	ws?: any;
+	ws?: WebSocket;
 	key?: Uint8Array;
 
-	promise1: any;
-	resolve1: any;
-	reject1: any;
-	promise2: any;
-	resolve2: any;
-	reject2: any;
-	promise3: any;
-	resolve3: any;
-	reject3: any;
+	state: State;
+	callback: (pc: RTCPeerConnection, newcode?: string) => void;
+	done: Promise<Uint8Array>;
+	resolve?: (fingerprint: Uint8Array) => void;
+	reject?: (reason: string) => void;
 
 	constructor(signalserver: string, code: string) {
+		this.signalserver = signalserver;
+		this.callback = () => {};
 		if (code !== "") {
 			[this.slot, this.pass] = webwormhole.decode(code);
 			if (this.pass.length === 0) {
 				throw "bad code";
 			}
 			console.log("dialling slot:", this.slot);
-			this.state = "b";
+			this.state = this.statePlayer2;
 		} else {
 			this.pass = crypto.getRandomValues(new Uint8Array(2));
 			console.log("requesting slot");
-			this.state = "a";
+			this.state = this.statePlayer1;
 		}
 
-		// There are 4 events that we need to synchronise with the caller on:
+		// There are 3 events that we need to synchronise with the caller on:
 		//   1. we got the first message from the signalling server.
 		//        We now have the slot number and the ICE server details, so we can
 		//        create the wormhole code and PeerConnection object, and pass them back
 		//        to the caller to display and configure, respectively.
 		//   2. the caller is done configuring the PeerConnection.
 		//        We can now create the offer or answer and send it to the peer.
 		//   3. we've successfully authenticated the other peer.
-		//        Signalling is now done, apart from any trickling candidates. The caller
+		//        Signalling is now done, apart from trickling candidates. The caller
 		//        can display the key fingerprint.
-		//   4. caller tells us the webrtc handshake is done. We can close the websocket.
-		this.promise1 = new Promise((resolve1, reject1) => {
-			this.promise2 = new Promise((resolve2, reject2) => {
-				this.promise3 = new Promise((resolve3, reject3) => {
-					this.resolve1 = resolve1;
-					this.reject1 = reject1;
-					this.resolve2 = resolve2;
-					this.reject2 = reject2;
-					this.resolve3 = resolve3;
-					this.reject3 = reject3;
-					this.dial(signalserver);
-				});
-			});
+
+		this.done = new Promise((resolve, reject) => {
+			this.resolve = resolve;
+			this.reject = reject;
 		});
 	}
 
-	async signal() {
-		return this.promise1;
+	async statePlayer1(data: string): Promise<State> {
+		const msg: { slot: string; iceServers: GoICEServers } = JSON.parse(data);
```

Comment:
Generally I prefer [type narrowing](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates) rather than faithfully casting from any.

Reply:
nice - i'll read up on this.